src/main/java/hudson/remoting/forward/CopyThread.java

```diff
@@ -16,10 +16,23 @@
     private final InputStream in;
     private final OutputStream out;
 
-    public CopyThread(String threadName, InputStream in, OutputStream out) {
+    public CopyThread(String threadName, InputStream in, OutputStream out, Runnable termination) {
+        this(threadName, in, out, termination, 0);
+    }
+
+    private CopyThread(String threadName, InputStream in, OutputStream out, Runnable termination, int previousTries) {
         super(threadName);
         this.in = in;
         this.out = out;
+        setUncaughtExceptionHandler((t, e) -> {
+            if (previousTries < 5) {
+                LOGGER.log(Level.FINE, "Uncaught exception in CopyThread " + t + ", retrying copy", e);
+                new CopyThread(threadName, in, out, termination, previousTries + 1).start();
```

Comment:
The problem that I see here is that the `finally` block in the previous thread would close the input and output streams, meaning that by the time it tries again, it can't use those streams. Perhaps the input and output stream closures should be moved elsewhere? Or is there a simpler solution that I'm not thinking of?

Reply:
Or just propagate the error to the caller thread as I've proposed above.

Reply:
I guess I'm confused about what should happen here.  Propagating the error is complicated - it's simply enough to pass the error into the termination handler, of course, but I don't know why the termination handler itself cares about the specific error.  From what I understand, and there's a good chance that my understanding is poor, is that the termination handler needs to clean up resources regardless of the specifics of the error.  Said cleanup will likely also cause other failures in the parent thread, as closed sockets can no longer be read from, etc.  Rethrowing the error in the parent thread is problematic at best because it would require waiting on a future (I think?) and ultimately doesn't seem worthwhile to me since the socket/stream closures will have no trouble causing errors themselves (which are in turn handled).  

What am I missing here?  (Yes, I've spent far too much time thinking about/writing random code around this.)